### PR TITLE
API: Infer extension types in array

### LINKED ISSUE #29791: Have `pd.array` infer new extension types
Currently `pd.array` sometimes requires an explicit `dtype=...` to get one of our extension arrays (we'll infer for Period, Datetime, and Interval).

This proposal is to have it infer the extension type for

* strings -> StringArray
* boolean -> BooleanArray
* integer -> IntegerArray

All of these currently return PandasArray.

Concretely, we'll need to teach `infer_dtype` how not to infer `mixed` for a mix of strings / booleans and NA values, similar to how it handles integer-na

```python
In [27]: lib.infer_dtype([True, None], skipna=False)
Out[27]: 'mixed'

In [28]: lib.infer_dtype(['a', None], skipna=False)
Out[28]: 'mixed'

In [29]: lib.infer_dtype([0, np.nan], skipna=False)
Out[29]: 'integer-na'
```

and then handle those in `array`.

### PR DESCRIPTION
**Pull request by TomAugspurger**
* string
* integer

Closes #29791 (though we'll want to add BooleanArray).

### CODE REVIEW

File: doc/source/user_guide/integer_na.rst [hunk 1]
```diff
@@ -50,17 +49,34 @@ NumPy array.
 You can also pass the list-like object to the :class:`Series` constructor
 with the dtype.
 
-.. ipython:: python
+.. warning::
 
-   s = pd.Series([1, 2, np.nan], dtype="Int64")
-   s
+   Currently :meth:`pandas.array` and :meth:`pandas.Series` use different
```
Comment (TomAugspurger) [inline doc/source/user_guide/integer_na.rst]: This is my attempt to explain the inconsistency between `pd.array` and `pd.Series` we're introducing. It's not ideal, but I think it's the right behavior for now.

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -1319,8 +1320,11 @@ def infer_dtype(value: object, skipna: object=None) -> str:
             return 'boolean'
 
     elif isinstance(val, str):
-        if is_string_array(values, skipna=skipna):
-            return 'string'
+        if is_string_array(values, skipna=True):
```
Comment (TomAugspurger) [inline pandas/_libs/lib.pyx]: It's unclear to me whether `mixed-string` is actually a useful distinction from just `string`. As an alternative, I can just always return `"string"` here, but that might break code relying on a mix of strings and NAs being "mixed".

File: pandas/core/construction.py [hunk 3]
```diff
@@ -246,21 +253,25 @@ def array(
...
     )
 
     if lib.is_scalar(data):
         msg = "Cannot pass scalar '{}' to 'pandas.array'."
         raise ValueError(msg.format(data))
 
-    data = extract_array(data, extract_numpy=True)
-
-    if dtype is None and isinstance(data, ABCExtensionArray):
+    if dtype is None and isinstance(
```
Comment (TomAugspurger) [inline pandas/core/construction.py]: The documentation stated that we use the `dtype` for Series, Index, and ExtensionArrays. Moved this up here to get the dtype from these before extracting.

This is important for the following:

```python
In [3]: s = pd.Series([1, 2])

In [4]: pd.array(s)
Out[4]:
<PandasArray>
[1, 2]
Length: 2, dtype: int64
```


Without this change, we would extract the data to an int64 ndarray. We ignore the dtype for ndarrays, so we would have `pd.array(pd.Series([1, 2]))` inferring `Int64Dtype()`, rather than NumPy's int64.

File: pandas/core/construction.py [hunk 4]
```diff
@@ -246,21 +253,25 @@ def array(
...
+        data, (ABCSeries, ABCIndexClass, ABCExtensionArray)
+    ):
         dtype = data.dtype
 
+    data = extract_array(data, extract_numpy=True)
```
Comment (jbrockmendel) [inline pandas/core/construction.py]: why is this moved below the dtype-extraction?

File: pandas/core/construction.py [hunk 5]
```diff
@@ -298,6 +309,12 @@ def array(
             # timedelta, timedelta64
             return TimedeltaArray._from_sequence(data, copy=copy)
 
+        elif inferred_dtype in {"string", "mixed-string"}:
+            return StringArray._from_sequence(data, copy=copy)
+
+        elif inferred_dtype in {"integer", "mixed-integer"}:
+            return IntegerArray._from_sequence(data, copy=copy)
```
Comment (jbrockmendel) [inline pandas/core/construction.py]: IIUC the IntegerArray part doesnt depend on the lib.inferr_dtype change?

Comment (TomAugspurger) [reply pandas/core/construction.py]: We already infer `[1, None]` as `mixed-integer`. The only change necessary was handling that case here, rather than falling through to PandasArray.

File: pandas/core/construction.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/construction.py]: See https://github.com/pandas-dev/pandas/pull/29799#discussion_r349765386

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (TomAugspurger) [reply pandas/_libs/lib.pyx]: I'm trying out just "string" now. Will see what break.s

File: pandas/core/construction.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/construction.py]: To repeat here what I just commented on the issue: shouldn't we be inferrring the dtype with `skipna=True`? Then [1, None] will also give 'integer' and not 'mixed-integer'

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/_libs/lib.pyx]: The same is true for `mixed-integer` no? (so we should be consistent?) But isn't that the purpose of passing `skipna=True` to not see missing values as non-strings?

**Comment by TomAugspurger** [comment]
CI is passing now.

File: doc/source/user_guide/integer_na.rst (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/user_guide/integer_na.rst]: Eventually we want to share code between these (and ideally also the Index constructor), right?

File: doc/source/whatsnew/v1.0.0.rst [hunk 6]
```diff
@@ -234,6 +234,37 @@ The following methods now also correctly output values for unobserved categories
 
    df.groupby(["cat_1", "cat_2"], observed=False)["value"].count()
 
+:meth:`pandas.array` inference changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:meth:`pandas.array` now infers pandas' new extension types in several cases (:issue:`29791`):
+
+1. Sting data (including missing values) now returns a :class:`arrays.StringArray`.
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.0.0.rst]: Sting -> String

File: pandas/core/construction.py [hunk 7]
```diff
@@ -298,6 +300,12 @@ def array(
             # timedelta, timedelta64
             return TimedeltaArray._from_sequence(data, copy=copy)
 
+        elif inferred_dtype == "string":
+            return StringArray._from_sequence(data, copy=copy)
+
+        elif inferred_dtype == "integer":
+            return IntegerArray._from_sequence(data, copy=copy)
+
         # TODO(BooleanArray): handle this type
```
Comment (jbrockmendel) [inline pandas/core/construction.py]: not necessarily for this PR, but is it viable to handle BooleanArray here now?

**Comment by jbrockmendel** [comment]
small comment about a typo, otherwise LGTM

**Comment by jorisvandenbossche** [comment]
Can you add BooleanArray as well?

File: doc/source/user_guide/integer_na.rst (hunk 1 above)
Comment (jorisvandenbossche) [reply doc/source/user_guide/integer_na.rst]: In the issue (https://github.com/pandas-dev/pandas/issues/29791), I mentioned for that reason the idea of a keyword to control this preference between old or new dtypes. 
(but we can also introduce that at the moment we want to share code if that seems useful then)

File: doc/source/whatsnew/v1.0.0.rst [hunk 8]
```diff
@@ -234,6 +234,37 @@ The following methods now also correctly output values for unobserved categories
...
+2. Integer data (including missing values) now returns a :class:`arrays.IntegerArray`.
+
+*pandas 0.25.x*
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: side issue, we are pretty inconsistent on showing the previous version in the whatsnew

**Review by jreback — APPROVED** [review]
lgtm. I think @jbrockmendel had some comments.

**Comment by jreback** [comment]
note that in #29791 should refer to teaching ``infer_dtype`` about extension types. This PR doesn't actually do that, rather you teach ``pd.array`` to correctly infer given the results of ``infer_dtype``. I actually think we *should* push this down and remove the logic from ``pd.array`` itself, BUT, I suspect that actually would break things (e.g. we would need to process 'nullable-integer') for example. So even though this is a positive change, I think we need to open an issue about this (or revise this PR).

**Comment by jbrockmendel** [comment]
LGTM

File: pandas/core/construction.py [hunk 9]
```diff
@@ -212,29 +221,20 @@ def array(
     [a, b, a]
     Categories (3, object): [a < b < c]
 
-    Because omitting the `dtype` passes the data through to NumPy,
-    a mixture of valid integers and NA will return a floating-point
-    NumPy array.
+    If pandas does not infer a dedicated extension type a
+    :class:`arrays.PandasArray` is returned.
```
Comment (jorisvandenbossche) [inline pandas/core/construction.py]: Should we mention that this can still change in the future? (eg that more types start to get inferred, so basically that you should not rely on the fact of `pd.array` returning a PandasArray when no dtype is specified)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
> #29791 should refer to teaching infer_dtype about extension types. This PR doesn't actually do that, rather you teach pd.array to correctly infer given the results of infer_dtype. I actually think we should push this down and remove the logic from pd.array

@jreback what would be the advantage of doing that? I personally don't really see the benefit of `infer_dtype` to be able to return both "integer" or "integer-nullable" for the same integer data. In the end, if you want to use the new dtype or a numpy dtype does not necessarily depend on the data, but from the context where it is being used (eg the context of `pd.array` wanting to return the new dtypes, so `pd.array` can decide how to interpret "integer", as numpy int or nullable int).

**Comment by jreback** [comment]
> > #29791 should refer to teaching infer_dtype about extension types. This PR doesn't actually do that, rather you teach pd.array to correctly infer given the results of infer_dtype. I actually think we should push this down and remove the logic from pd.array
> 
> @jreback what would be the advantage of doing that? I personally don't really see the benefit of `infer_dtype` to be able to return both "integer" or "integer-nullable" for the same integer data. In the end, if you want to use the new dtype or a numpy dtype does not necessarily depend on the data, but from the context where it is being used (eg the context of `pd.array` wanting to return the new dtypes, so `pd.array` can decide how to interpret "integer", as numpy int or nullable int).

you are missing the point
this PR is just a temporary work around - infer_dtype is used practically everywhere 
this is a very specific case that you are trying to solve

**Comment by jorisvandenbossche** [comment]
Jeff, if you think I am missing your point, then try to explain it better (but I rather think we are just disagreeing). 
In any case, it's not a discussion for this PR I think, so maybe you can open a new issue for this?

File: pandas/core/construction.py [hunk 10]
```diff
@@ -270,7 +275,7 @@ def array(
         return cls._from_sequence(data, dtype=dtype, copy=copy)
 
     if dtype is None:
-        inferred_dtype = lib.infer_dtype(data, skipna=False)
+        inferred_dtype = lib.infer_dtype(data, skipna=True)
```
Comment (jreback) [inline pandas/core/construction.py]: so my issue with this PR is that is duplicating a lot of logic that is already held here: https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/lib.pyx#L1948, so now we have 2 places with slightly different ways of doing things.

This routine is slightly more 'high-level', but myabe_convert_objects is way more used internally. So how to reconcile these things?

@jorisvandenbossche @TomAugspurger

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/construction.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/core/construction.py]: I'm not familiar with `maybe_convert_objects`, but what's the duplicate logic between the two? At a glance, it seems like `maybe_convert_objects` is mixing two things

1. type inference (the sole purpose of `lib.infer_dtype`)
2. array construction (which may be `pd.array`? Or move the core to some `_libs` method?).

Should we update `maybe_infer_objects` to use `array` internally? I don't have a feel for whether that's even possible, do you?

**Comment by TomAugspurger** [comment]
Fixed the docstring & merging master. I can open an issue about deduplicating `array`, `maybe_convert_object` and `infer_dtype`. I think that can be a followup though, as this PR really is only changing ~10 LOC, despite the large diff :)

**Comment by jreback** [comment]
> Fixed the docstring & merging master. I can open an issue about deduplicating `array`, `maybe_convert_object` and `infer_dtype`. I think that can be a followup though, as this PR really is only changing ~10 LOC, despite the large diff :)

right not trying to cause an issue :->

just wanting to clarify the apis & implementation we have for internal (maybe_convert_*) and external (pd.array) that need to be cleanly separated and useful / documented etc.

opening an issue would be great.

**Review by jreback — APPROVED** [review]
lgtm. ping on green.,

**Comment by TomAugspurger** [comment]
All green.

File: pandas/core/construction.py (hunk 10 above)
Comment (jbrockmendel) [reply pandas/core/construction.py]: > Should we update maybe_infer_objects to use array internally? I don't have a feel for whether that's even possible, do you?

Probably not in the short term.  Best guess for de-duplication for the discussed functions will look something like:

- `lib.maybe_convert_objects` is made to back `lib.infer_dtype`
  - Note: `lib.maybe_convert_objects` involves two runtime non-cython imports that I'd really like to find a way to avoid (one will be easy, the other very much not)
- `pd.array` calls `maybe_convert_objects` instead of `infer_dtype`
  - More generally, many places where we call infer_dtype followed by casting can be replaced to be one-pass instead of two-pass.
- `Series` constructor is backed by `pd.array`
- `Index.__new__` is backed by `pd.array` (we'd need something like `RangeArray` first)

**Comment by jreback** [comment]
thanks @TomAugspurger

**Comment by TomAugspurger** [comment]
Thanks!
